rfcs/20190208-pybind11.md

```diff
@@ -0,0 +1,327 @@
+    )
+
+    native.genrule(
+        name = name + "_version_script",
+        outs = [version_script_file],
+        cmd = "echo '{global:\n %s;\n %s;\n %s;\n local: *;};' >$@" % (symbol, symbol2, symbol3),
+        output_licenses = ["unencumbered"],
+        visibility = ["//visibility:private"],
+        testonly = testonly,
+    )
+    native.cc_binary(
+        name = so_file,
+        srcs = srcs + hdrs,
+        data = data,
+        copts = copts,
+        nocopts = nocopts,
+        linkopts = linkopts + select({
+            "@org_tensorflow//tensorflow:windows": [],
+            "@org_tensorflow//tensorflow:darwin": [
+                "-Wl,-exported_symbols_list",
+                exported_symbols_file,
+            ],
+            "//conditions:default": [
+                "-Wl,--version-script",
+                "$(location %s)" % version_script_file,
+            ],
+        }),
+        deps = deps + [
+            exported_symbols_file,
+            version_script_file,
+        ],
+        features = features,
+        linkshared = 1,
+        testonly = testonly,
+        licenses = licenses,
+        visibility = visibility,
+        deprecation = deprecation,
+        restricted_to = restricted_to,
+        compatible_with = compatible_with,
+    )
+    native.genrule(
+        name = name + "_pyd_copy",
+        srcs = [so_file],
+        outs = [pyd_file],
+        cmd = "cp $< $@",
+        output_to_bindir = True,
+        visibility = visibility,
+        deprecation = deprecation,
+        restricted_to = restricted_to,
+        compatible_with = compatible_with,
+    )
+    native.py_library(
+        name = name,
+        data = select({
+            "@org_tensorflow//tensorflow:windows": [pyd_file],
+            "//conditions:default": [so_file],
+        }),
+        srcs_version = srcs_version,
+        licenses = licenses,
+        testonly = testonly,
+        visibility = visibility,
+        deprecation = deprecation,
+        restricted_to = restricted_to,
+        compatible_with = compatible_with,
+    )
+```
+
+### Issues
+
+
+
+1.  PyObject pointers are not inherent to pybind11 and require some minor code modification. [[GitHub Issue Link](https://github.com/pybind/pybind11/issues/1201#issue-278504106)]. As you notice in diagram 2 of the rollout we have to create a small wrapper function around the function. Pybind11 cannot cast a python object to `PyObject*` so you need to provide a `pybind11::handle` reference and cast it internally. We could potentially modify the IsSequence function directly, but that would require additional unnecessary modification in the .cc file. 
+2.  Complicated use cases where we export weird classes are not properly documented or seen before. We unfortunately don't know what we don't know when it comes to our migrations. [Other users](https://community.lsst.org/t/using-pybind11-instead-of-swig-to-wrap-c-code/1096) have also complained about the more complicated use cases.
```

Comment:
Could you give an example of a problematic case in TensorFlow? Also note that the LSST post could be out of date (published in 2016), e.g. `pybind11` does support sequence types now via `pybind11::sequence`.

Reply:
Yes you are right, that post is out of date. Fortunately from the minor conversions I've made I haven't seen any roadblocking issues. This was a more generic comment that could impact the roll-out timeline.